### ISSUE TITLE
Device > Channels > Import > The "Task finished" snackbar is never shown

### ISSUE THREAD

**Issue by pcenov** [learningequality/kolibri#12965]
## Observed behavior

This is a follow-up to https://github.com/learningequality/kolibri/pull/12946 - The "Task finished" snackbar is never shown after successfully importing/exporting or deleting a channel. This is valid for both Kolibri 0.17 and 0.18.

## Expected behavior

Should be investigated why is this snackbar not being displayed and if possible it should be restored.

## Steps to reproduce the issue

1. Install the latest develop build or Kolibri 0.17.4
2. Go to Device > Channels > Import and import/export/delete a channel
3. Observe that only the "Task started" snackbar is displayed

## Usage Details
Windows 11, Ubuntu 22  - Chrome

**Comment by pcenov** [comment]
@rtibbles

**Comment by rtibbles** [comment]
Thanks @pcenov

**Comment by shruti862** [comment]
@rtibbles @pcenov Is this issue open to work upon??

**Comment by rtibbles** [comment]
Yes - ensuring that the task finished snackbar properly displays when a task finishes would be what is needed!

**Comment by shruti862** [comment]
@rtibbles Could you please assign the issue to me? I would like to give it a try.

**Comment by ananyakaligal** [comment]
@rtibbles @pcenov I would also like to give it a try, could you please assign me this?

**Comment by LianaHarris360** [comment]
Thank you both for your willingness to help with this! We only need one person to handle it at a time, and since @shruti862  asked first, I'll assign it to them.

**Comment by karankoder** [comment]
Hi @LianaHarris360 , I noticed that this issue is still open, and it seems @shruti862 is not actively working on it. Could you please assign this issue to me? I'd like to take it up

**Comment by LianaHarris360** [comment]
Hi @karankoder, we'll wait another week since the holidays are coming to an end. If there's no update by then, I'll hand it over to @ananyakaligal if they are still interested, if not I'll assign to you. @shruti862, can you let me know if you will be working on this?

**Comment by shruti862** [comment]
@LianaHarris360 Yes, I am working on it; I have already inspected the cause of issue but due to the holidays I thought to share afterwards ,By tonight I will share my work.

**Comment by LianaHarris360** [comment]
Thanks for the update and your thoughtfulness in waiting until after the holidays to share @shruti862! It's great that you're making progress, there's no rush to share it by tonight :)

**Comment by shruti862** [comment]
@rtibbles @pcenov ,After inspecting the issue I noticed that watcher on computed property (watchedTaskHasFinished ) is not being triggered in taskNotificationMixin.js file due to which createTaskFinishedSnackbar() method is not called leading to no display of finished Snackbar. But I was unable to inspect the cause of watcher not being triggered because both watchedTaskId and watchedTaskHasFinished values are changing on task starting and on ending..Could you please suggest me something which can help me figure out the problem 
```
computed: {
    watchedTaskId() {
      return this.$store.state.manageContent.watchedTaskId;
    },
    watchedTaskHasFinished() {
      // Should switch between null and the taskIds, and can be watched
      // by component to trigger side effects when a task finishes.
      return this.$store.getters['manageContent/taskFinished'](this.watchedTaskId);
    },
  },
  watch: {
    watchedTaskHasFinished(val) {
        if (val && val === this.watchedTaskId) {
          if (this.showSnackbarWhenTaskHasFinished) {
            this.createTaskFinishedSnackbar();
          }
          if (typeof this.onWatchedTaskFinished === 'function') {
            this.onWatchedTaskFinished();
          }
        }
     
    },
  },
```

**Comment by rtibbles** [comment]
Hrm, that is tricky - and this watcher is never called at all?

```
  watch: {
    watchedTaskHasFinished(val) {
        if (val && val === this.watchedTaskId) {
          if (this.showSnackbarWhenTaskHasFinished) {
            this.createTaskFinishedSnackbar();
          }
          if (typeof this.onWatchedTaskFinished === 'function') {
            this.onWatchedTaskFinished();
          }
        }
     
    },
  },
```

Or is it called only with a falsy value so it doesn't get past the `if` statement? Or a value that isn't equal to `this.watchedTaskId`?

**Comment by shruti862** [comment]
@rtibbles, According to my finding watchers is never called, But when I tried to call it on initial render by using immediate:true , it does triggered because at that time it is not dependent on change in a computed property.
 watchedTaskId computed property is working fine ,I think the issue resides in  taskFinished getter which is being called in  watchedTaskHasFinished computed property. But again unable to figure it out .

**Comment by AlexVelezLl** [comment]
Hey @shruti862! Are we sure that this taskNotificationMixin is being used in that page? I just added a console.log on mounted, but didnt see any outputs in the console. Can you double check if that mixin is used in the manage tasks page?

**Comment by shruti862** [comment]
Hey @AlexVelezLl  After inspecting using vue devtool I found that in manageChannelContentPage, on clicking delete button handleDeleteSubmit method evoke which has this.onTaskSuccess method which displays 'Task started' Snackbar through notifyAndWatchTask(task)
```
 handleDeleteSubmit({ deleteEverywhere }) {
        this.beforeTask();
        this.startDeleteTask({
          deleteEverywhere,
          channelId: this.channelId,
          channelName: this.channel.name,
          included: this.selections.included,
          excluded: this.selections.excluded,
        }).then(this.onTaskSuccess, this.onTaskFailure);
      },
      handleExportSubmit({ driveId }) {
        this.beforeTask();
        this.startExportTask({
          driveId,
          channelId: this.channelId,
          channelName: this.channel.name,
          included: this.selections.included,
          excluded: this.selections.excluded,
        }).then(this.onTaskSuccess, this.onTaskFailure);
      },
      beforeTask() {
        this.closeModal();
        this.bottomBarDisabled = true;
      },
      onTaskSuccess(task) {
        this.bottomBarDisabled = false;
        this.watchedTaskType = task.type;
        this.notifyAndWatchTask(task);
      },
```
and it also have this  onWatchedTaskFinished() method which is supposed to be used by taskNotificationMixin to display 'Task Finished' Snackbar
```
/**
       * @public
       * Used by the taskNotificationMixin to handle the completion of the task
       */
      onWatchedTaskFinished() {
        // clear out the nodeCache
        this.nodeCache = {};
        // clear out selections
        this.$store.commit('manageContent/wizard/RESET_NODE_LISTS');
        // refresh the data on this page. if the entire channel ends up
        // being deleted, then redirect to upper page
        this.fetchPageData(this.channelId)
          .then(this.setUpPage)
          .catch(error => {
            // If entire channel is deleted, redirect
            if (error.response.status === 404) {
              this.$router.replace({ name: PageNames.MANAGE_CONTENT_PAGE });
            } else {
              this.$store.dispatch('handleApiError', { error });
            }
          });
      },
```

**Comment by shruti862** [comment]
@AlexVelezLl I found nothing revelant  in ManageTaskPage which is responsible for displaying Snackbar

**Comment by rtibbles** [comment]
It seems reasonable that there may be an issue in the reactivity of `taskFinished` getter, based on your observations. My guess would be is that it is not updating when the underlying vuex state updates.

I would look and see if watching either the task itself, or the task's status triggers a watcher? Then we can just check the status in the watcher to see if it has finished and trigger the snackbar based on that?

**Comment by shruti862** [comment]
Hey @rtibbles ,When I tried to have tasks computed property, it is not updating with finished tasks, it remains a empty array is it due to reactivity issues? In Vuex store taskList is being updated but not reflecting changes in tasks computed property
```
computed: {
    watchedTaskId() {
      return this.$store.state.manageContent.watchedTaskId;
    },
    tasks() {
      return this.$store.state.manageContent.taskList;
    },
  },
```

**Comment by shruti862** [comment]
Hey @rtibbles , any suggestions on resolving this reactivity issue would be appreciated so I can start working on it. If there are no updates or if someone else is better suited to take it up, feel free to unassign me so others can contribute. Thanks!

**Comment by rtibbles** [comment]
So, it seems the way that the taskList vuex state is being updated is not triggering reactivity. What is the vuex mutation that is updating it? Where is it being updated?

**Comment by shruti862** [comment]
Hey @rtibbles , in <code>kolibri/plugins/device/assets/src/modules/manageContent/index.js </code>file I found mutation that is updating taskList
```
function defaultState() {
  return {
    channelList: [],
    channelListLoading: false,
    taskList: [],
    watchedTaskId: null,
  };
}

export default {
  namespaced: true,
  state: defaultState,
  mutations: {
    SET_STATE(state, payload) {
      Object.assign(state, payload);
    },
    RESET_STATE(state) {
      Object.assign(state, defaultState());
    },
    SET_CHANNEL_LIST(state, channelList) {
      state.channelList = [...channelList];
    },
    SET_CHANNEL_LIST_LOADING(state, isLoading) {
      state.channelListLoading = isLoading;
    },
    SET_TASK_LIST(state, taskList) {
      state.taskList = [...taskList];
    },
    SET_WATCHED_TASK_ID(state, taskId) {
      state.watchedTaskId = taskId;
    },
```

**Comment by rtibbles** [comment]
Hrm, that does look like it shouldn't be falling foul of any of the usual reactive gotchas.

Can you try using the `mapState` helper function from vuex to define the `tasks` computed property instead of directly accessing the store to see if that makes a difference?

See here for guidance: https://vuex.vuejs.org/guide/state.html#the-mapstate-helper (looking at how it is used elsewhere in Kolibri will also help!)

**Comment by shruti862** [comment]
Hey @rtibbles, Thank you for your suggestion. I will try this approach as well and provide you with updates soon.

**Comment by shruti862** [comment]
Hey @rtibbles, I tried your suggestion 
Code changes:
```
 computed: {
    ...mapState('manageContent', {
      taskList: state => state.taskList,
      watchedTaskId: state => state.watchedTaskId
    }),
  },
```
Before deleting resources computed properties were:

![Image](https://github.com/user-attachments/assets/ca828105-8718-4f97-b112-05f7ccda1313)

After deleting resources, only watchedTaskId is  getting updated ( as seen under vuex binding in vue-devtools ),taskList still remains empty array:

![Image](https://github.com/user-attachments/assets/b6eb2c72-2223-45ad-91b4-17abbe6d9a35)

**Comment by MisRob** [comment]
Hi @shruti862, would you be able to debug a bit deeper around what's the reactivity issue around `taskList`? Maybe there is a place in code that doesn't update the array in a way that breaks the reactivity chain? https://v2.vuejs.org/v2/guide/reactivity.html#Change-Detection-Caveats

**Comment by MisRob** [comment]
If this doesn't help then I'd recommend you track step by step places executed in code that (1) set and (2) retrieve the task list and see at what point exactly the mismatch begins

**Comment by shruti862** [comment]
@MisRob Thank you for your suggestions I will try to debug it deeper and will provide updates soon.

**Comment by Yashpreet-Singh-Jagdev** [comment]
Hey @shruti862  are you still working on this issue? If not then maybe @LianaHarris360 can reassign me this issue. I'd like to take it up and begin my contributions in Kolibri !

**Comment by shruti862** [comment]
Hey @Yashpreet-Singh-Jagdev, I am currently occupied with other issues, so this one is on hold for now. However, since I have already spent time on it in the past, I would love to resolve it. I’ll be getting back to it in a few days.

**Comment by shruti862** [comment]
Hey @LianaHarris360 , I would like to unassign myself from this issue as I am currently busy with other engagements. Also, I see that many new contributors are looking for issues to work on, so I’d like to free this up for them. Thanks!

**Comment by LianaHarris360** [comment]
@Yashpreet-Singh-Jagdev If you are still interested, I can assign this issue to you?

**Comment by Yashpreet-Singh-Jagdev** [comment]
Sure, I would love to work on this issue.

**Comment by yashhash2** [comment]
Hi @Yashpreet-Singh-Jagdev are you still working on this issue if not then @LianaHarris360 can assign it to me if you're still working then all the best :)

**Comment by Yashpreet-Singh-Jagdev** [comment]
Hey @yashhash2 as discussed  [here](https://github.com/learningequality/kolibri/issues/12198) since that issue is almost about to get completed I will now focus on this issue. Although,  thanks for asking.

**Comment by Sahil-Sinha-11** [comment]
Hey @LianaHarris360, I was wondering if @Yashpreet-Singh-Jagdev isn't currently working on this issue—would it be alright if I gave it a try?

**Comment by LianaHarris360** [comment]
Yes, I'll go ahead and assign you to this issue.

**Comment by Sahil-Sinha-11** [comment]
Hey, I found that the taskFinished (state) is not working. When I logged taskList and ids, it came in the console, but nothing inside the taskFinished() function logs to the console.

<img width="603" alt="Image" src="https://github.com/user-attachments/assets/5426cb49-1ed1-4cc1-997f-88a5fb532aaf" />

<img width="603" alt="Image" src="https://github.com/user-attachments/assets/f623b27b-b8b5-4750-915e-786d8da1bc82" />

**Comment by MisRob** [comment]
Hi @Sahil-Sinha-11, that sounds like you're on the right track. Do you want to continue this task and keep debugging to get to the root of the problem?

**Comment by Sahil-Sinha-11** [comment]
Yes @MisRob, I am on it.

**Comment by Sahil-Sinha-11** [comment]
Hey, on logging the Task list, the task gets updated, but watchedTaskId is being cleared too early (before completion check)
Tasks aren't being properly marked as clearable: true when complete. For this, I tried to delay the on watch task finished function and called createFinishedTaskbar in onWatchedTaskFinished, but it didn't work, can you guide me here?

<img width="550" alt="Image" src="https://github.com/user-attachments/assets/eaff81e4-2687-4fcf-8c54-2749a34b076d" />

<img width="563" alt="Image" src="https://github.com/user-attachments/assets/d01a4142-c9ce-43e2-8875-8c9ce4a09771" />

**Comment by MisRob** [comment]
Hi@Sahil-Sinha-11, thanks for looking into it. This sound like some kind of race condition? Those can be tricky. I think you'd need to continue debugging to map exactly how this is happening. If it's difficult to make progress, choosing another issue instead of this one could be more suitable.

**Comment by Sahil-Sinha-11** [comment]
Hello @MisRob, please feel free to unassign me from the issue. However, I would still like to continue contributing to it in the background. 🙂

**Comment by AlexVelezLl** [comment]
Thanks for letting us know @Sahil-Sinha-11. I will unassign this from you. But don't hesitate to ask to be reassigned if you want to give it another try or find something new :)
